### PR TITLE
Api/client/base api service

### PR DESCRIPTION
**Pull request by ElijahRW**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
    Added Simple API service to Angular client. _(This will be refactored going forward. but Functionality should affect calling services heavily.)_

### CODE REVIEW

File: src/app/services/server-api.service.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+@Injectable()
+export class ServerApiService {
+  // TODO: Configure base Url, to point to hosted endpoint. (localhost:3333 isn't a valid endpoint.)
+  private baseUrl = 'https://jsonplaceholder.typicode.com/todos/1';  // URL to web api
+
+  constructor(private http: HttpClient) { }
+
+  // Base Function
+  public getFromDataBase(urlExtension: string): Observable<any> {
```
Comment (gstone271) [inline src/app/services/server-api.service.ts]: I'd drop the Smurf Naming Convention and just call these `get`, `post`, `delete`.
I think it would be useful to have the query parameters (`?foo=bar,frob=baz`) represented as an object passed to this function (`{foo: 'bar', frob: 'baz'}`), so that callers don't need to know how to encode URLs

File: src/app/services/server-api.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+describe('AppointmentsService', () => {
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      providers: [ServerApiService]
+    });
+  });
+
+  it('should be created', inject([ServerApiService], (service: ServerApiService) => {
+    expect(service).toBeTruthy();
+  }));
```
Comment (gstone271) [inline src/app/services/server-api.service.spec.ts:14]: Can you add an actual test case showing that the class works?  Eg. calling the Hello World API. Ofc we'll need to change it later once we have an actual API to call.

File: docs/client-api-instrucitons.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+### To Create a Service using the Client's Base API 
+  * Create the new desired service. (This should be housed in the same folder as the component being serviced)
+  * Include the Api service: `import { ServerApiService } from '../services/ServerApi.service';`
+  * Add the Service to the Component Constructor: `constructor (private serverApiService: ServerApiService) { }`
+  * Call the GET/POST/DELETE Service as desired. (Add the additional data to the url as desired).
+    * This function will return an `Observable<any>` object to be used as deemed fit. The calling function should typecheck this object by returning a typed observable object.
+    * Example: 
+    ```javascript
+        public getAppointmentReference(): Observable<AppointmentReference> {
+            return this.app(this.ServerApiService.getFromDatabase('appointments?id=1'));
```
Comment (gstone271) [inline docs/client-api-instrucitons.md]: Is `app` a real function? If so what does it do? If not, it should be cut from this example.

File: src/assets/dbObjects/example.json [hunk 4]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (gstone271) [inline src/assets/dbObjects/example.json]: Where is this file used?

**Review by gstone271 — CHANGES_REQUESTED** [review]
Seems like a good use of a Service; some questions about details

File: src/assets/dbObjects/example.json (hunk 4 above)
Comment (ElijahRW) [reply src/assets/dbObjects/example.json]: No where, I forgot that json file made it in.

File: src/app/services/server-api.service.spec.ts (hunk 2 above)
Comment (ElijahRW) [reply src/app/services/server-api.service.spec.ts:14]: Sure thing. Do you mean that I call an external API (Such as `http://dummy.restapiexample.com/` to confirm that the Service actually reaches the web?)

File: src/app/services/server-api.service.ts (hunk 1 above)
Comment (ElijahRW) [reply src/app/services/server-api.service.ts]: Understood, How would I go about adding query parameters without knowing how many parameters are being passed in?

File: src/app/services/server-api.service.spec.ts (hunk 2 above)
Comment (ElijahRW) [reply src/app/services/server-api.service.spec.ts:14]: And should this be a mocked test? Or a function within server-api.service.ts?

**Review by Liriaene — APPROVED** [review]
LGTM. But front end isn't my strong point so idk if it's a valid approval

File: src/app/services/server-api.service.ts (hunk 1 above)
Comment (gstone271) [reply src/app/services/server-api.service.ts]: You accept one argument, a Map or Object, and iterate over its keys. Use https://angular.io/guide/http#url-parameters , but don't accept an HttpParams argument; instead construct an HttpParams based on the object you receive.

File: src/app/services/server-api.service.spec.ts (hunk 2 above)
Comment (gstone271) [reply src/app/services/server-api.service.spec.ts:14]: I dunno how the existing tests work, I just want the code to actually be run somehow, and to be able to repeat that run. Code that has never been run is strictly theoretical code.

File: src/app/services/server-api.service.ts (hunk 1 above)
Comment (gstone271) [reply src/app/services/server-api.service.ts]: Still needs addressing

File: docs/client-api-instrucitons.md (hunk 3 above)
Comment (gstone271) [reply docs/client-api-instrucitons.md]: Still wondering about this

**Review by gstone271 — CHANGES_REQUESTED** [review]
earlier troubles still unresolved

**Review by wynnliam — COMMENTED** [review]
This seems pretty good and informative to me. All my issues with it are the same ones Griffin pointed out, but otherwise this is solid work.

File: src/app/services/server-api.service.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    return '15';
+  }
+
+  /*
+  * Handle Http operation that failed.
+  * Let the app continue.
+  * @param operation - name of the operation that failed
+  * @param result - optional value to return as the observable result
+  */
+  private handleError<T>(operation = 'operation', result?: T) {
```
Comment (gstone271) [inline src/app/services/server-api.service.ts]: `result` is always null.

File: src/app/services/server-api.service.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    return (error: any): Observable<T> => {
+
+      // TODO: send the error to remote logging infrastructure
+      console.error(error); // log to console instead
+
+      // TODO: better job of transforming error for user consumption
+      console.log(`${operation} failed: ${error.message}`);
+
+      // Let the app keep running by returning an empty result.
+      return of(result as T);
```
Comment (gstone271) [inline src/app/services/server-api.service.ts]: I'd prefer this was written `observable.of` if possible.

File: src/app/services/server-api.service.ts (hunk 5 above)
Comment (gstone271) [inline src/app/services/server-api.service.ts]: There needs to be some way for the caller to decide how to handle errors. Eg if the user mistyped password, it will produce an error; caller needs to detect that it was an AuthenticationExpired error.

File: src/environments/environment.ts [hunk 7]
```diff
@@ -67,3 +67,9 @@ export const decorateModuleRef = _decorateModuleRef;
 export const ENV_PROVIDERS = [
   ...PROVIDERS
 ];
+
+export const environment = {
+  production: false,
```
Comment (gstone271) [inline src/environments/environment.ts]: Why is this needed?

File: src/environments/environment.ts [hunk 8]
```diff
@@ -67,3 +67,9 @@ export const decorateModuleRef = _decorateModuleRef;
...
+
+  apiUrl: 'http://localhost:3000'
```
Comment (gstone271) [inline src/environments/environment.ts]: This seems unused.

File: src/app/customer-search/customer-search.service.ts [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+import { Observable } from 'rxjs';
+
+@Injectable()
+export class CustomerSearchService {
+
+    constructor(private service: ServerApiService) { }
+
+    public searchApiService(PatientId: string): Observable<Response<Appointment[]>> {
+        console.log('default send');
```
Comment (gstone271) [inline src/app/customer-search/customer-search.service.ts]: Message seems useless.

Comment (ElijahRW) [reply src/app/customer-search/customer-search.service.ts]: Oops... Yes... yes it is

File: src/environments/environment.ts (hunk 7 above)
Comment (ElijahRW) [reply src/environments/environment.ts]: This was somehow included in the Rebase. I'll remove it.

File: src/environments/environment.ts (hunk 8 above)
Comment (ElijahRW) [reply src/environments/environment.ts]: It is, I'll revert changes to environment.ts

File: src/app/services/server-api.service.ts (hunk 5 above)
Comment (ElijahRW) [reply src/app/services/server-api.service.ts]: Would a function as a variable be a good solution for this? (IE, how you have the error handling configured for the API?

File: src/app/services/server-api.service.spec.ts (hunk 2 above)
Comment (ElijahRW) [reply src/app/services/server-api.service.spec.ts:14]: I've added the Appointment service that the code effectively runs. I plan to add tests. However, I think test functions are something we can add in during cleanup week.

**Review by gstone271 — APPROVED** [review]
Haven't tested CustomerSearchComponent but the service looks fine to unblock folks, we can figure out error handling once happy path is working.

File: api-objects/Appointment.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
+export class Appointment {
+    public appointmentid: number;
```
Comment (Liriaene) [inline api-objects/Appointment.ts:2]: I don't think appointment Id is something that is sent to the API ever because it's auto generated. 
If this is for the client so the client can recieve an appointment id then may we should consider returning tfi, ths, ans ts id as well?

File: api-objects/Appointment.ts [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    public tfi_r: number;
+    public tfi_q: number;
+    public tfi_e: number;
+    public tfi_overallscore: number;
+    public ths_sectiona: number;
+    public ths_sectionb: number;
+    public ths_sectionc: number;
+    public ths_sectionc_example: string;
+    public ts_type: string;
+}
```
Comment (Liriaene) [inline api-objects/Appointment.ts]: Obligatory newline comment

File: src/app/customer-search/customer-search.component.ts [hunk 12]
```diff
@@ -3,92 +3,80 @@ import { TestsDataService } from '../services/tests-data.service';
...
-  public nextPage(amt: number) {
-    while ((this.currentPage + 1) * 10 < this.resultsTable.length && amt > 0) {
-      this.currentPage++;
-      amt--;
+    // CHANGE this function to actually load the selected appointment into sessionStorage and tell
+    // audiologist-navigation to change state
+    public loadAppt(appt: Appointment) {
+        if (appt['id'] === '') { return; }
+        // for each dataService saveData
+        console.log('appt: ID-"' + appt.appointmentid + '"      Audiologist-"' + appt.authorityname + '"');
```
Comment (Liriaene) [inline src/app/customer-search/customer-search.component.ts:44]: Unsure why the client needs to log to console, If this is for logging purposes, perhaps there's a better way?

**Review by Liriaene — APPROVED** [review]
Overall, everything looks good. My only concern is that there is a stray console.log that I'm unsure is needed but it doesn't look like it effects functionality.
Also some obligatory newline at end of file comments.

File: src/app/customer-search/customer-search.component.ts (hunk 12 above)
Comment (ElijahRW) [reply src/app/customer-search/customer-search.component.ts:44]: This is a placeholder console call to verify that the Appointment details are properly held. It will be replaced once the Appointment search is connected to the Appointment summary components are connected

File: api-objects/Appointment.ts (hunk 10 above)
Comment (ElijahRW) [reply api-objects/Appointment.ts:2]: This is what is received by the client from the Server. Appointment ID is very useful if we need to update notes or make other changes.
